docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -313,34 +532,124 @@ type LocalObjectTemplate struct {
+     path: “/spec/template/spec/kubeadmConfigSpec/files”
+     valueFrom:
+       template: |
+         - contentFrom:
+             secret
+               key: control-plane-azure.json
+               name: {{ .builtin.cluster.name }}-control-plane-azure-json
+   ```
 
 #### Validations
+
 ##### ClusterClass
 - For object creation:
   - (defaulting) if namespace field is empty for a reference, default it to `metadata.Namespace`
   - all the reference must be in the same namespace of `metadata.Namespace`
   - `spec.workers.machineDeployments[i].class` field must be unique within a ClusterClass.
+  - `VariableDefinitionClass`:
+    - names must be unique, not empty and not equal to `builtin`
+    - schemas must be valid
+  - `PatchClass`:
+    - names must be unique and not empty
+  - `PatchDefinition`:
+    - selector must have valid field values and match at least one template
+    - jsonPatches:
+      - `op`: one of: `add`, `replace` or `remove`
+      - `path`:
+        - must be a valid JSON pointer (RFC 6901) and start with `/spec/`
+        - only the indexes `0` (prepend) and `-` (append) in combination with the `add` operation are allowed (append and prepend 
+          are the only allowed array modifications).
+        - the JSON pointer is not verified against the target CRD as that would require parsing the template CRD, which is impractical in a webhook.
+      - either `value` or `valueFrom` have to be set when op is `add` or `replace`
+        - `valueFrom`:
+          - either `variable` or `template` have to be set
+          - `variable` (if set): should be a valid variable
+          - `template` (if set): should be a valid Go template
+            - We consider validating the Go template by generating values for all variables used in the template, then evaluating 
+              the template and validating if the result is a valid YAML or JSON value.
+    - We will only implement syntax validation. A semantic validation is not possible as the patches will be applied 
+      to provider templates. **It’s the responsibility of the ClusterClass author to ensure the patches are semantically valid**.
+
 - For object updates:
   - all the reference must be in the same namespace of `metadata.Namespace`
   - `spec.workers.machineDeployments[i].class` field must be unique within a ClusterClass.
   - `spec.workers.machineDeployments` supports adding new deployment classes.
+  - `VariableDefinitionClass`:
+    - names must be unique, not empty and not equal to `builtin`
+    - schemas must be valid
+    - schemas are mutable
+      - The current assumption is that we validate schema changes against existing clusters and block in case the changes are 
+        not compatible. We are considering to implement a clusterctl command to allow the user to simulate the impact of a schema change, 
+        because the validation might be too complex for a webhook (but this is still TBD). This command could also be used to simulate patches.
```

Comment:
A similar command to "render" clusters from a class would also be very interesting for creation of classes and CI. That would also allow further validation of patches.

Reply:
It's not part of it already imo. That command is only about creating new clusters. I think it would be very useful to be able to statically render a ClusterClass, ideally by just providing all the relevant files. That would make development of ClusterClasses (especially patches) and so on much easier.

And since I have a GitOps perspective: It would allow to statically validate ClusterClasses in a CI pipeline.

Reply:
Just to confirm. So the command should be able to read a local file with the cluster class and one with a cluster and then output the resulting resources of the reconciliation?

Reply:
I don't think changes are required because IMO this falls under 

> We are considering to implement a clusterctl command to allow the user to simulate the impact of a schema change

And this thread is a +1 to this idea.
Eventually, let's add "; this will be also useful for validation of changes in GitOps pipelines." to keep explicit track of this.

Reply:
@schrej I think https://github.com/kubernetes-sigs/cluster-api/issues/5319 will address this use case. Feel free to comment on #5319 in case there's something missing

Reply:
I agree. Sorry that I didn't respond sooner, I was on vacation the past two weeks.

Reply:
No problem :). I'll resolve